backend/graph/messages.go

```diff
@@ -0,0 +1,62 @@
+package graph
+
+import (
+	"context"
+	cache "github.com/Code-Hex/go-generics-cache"
+	"github.com/ansel1/merry/v2"
+	"github.com/bcc-code/brunstadtv/backend/common"
+	"sync"
+	"time"
+)
+
+type timedCacheEntry struct {
+	Cached time.Time
+	Entry  any
+}
+
+var messageCache = cache.New[string, timedCacheEntry]()
+
+var messageLock = sync.Mutex{}
+
+func getMessagesFromCache(timestamp *time.Time) ([]common.MaintenanceMessage, bool) {
+	if messages, ok := messageCache.Get("maintenance"); ok && (timestamp == nil || messages.Cached.Equal(*timestamp) || messages.Cached.After(*timestamp)) {
+		return messages.Entry.([]common.MaintenanceMessage), true
+	}
+	return nil, false
+}
+
+var truncateTimeInterval time.Duration = 1
+
+func (r *Resolver) getMaintenanceMessages(ctx context.Context, timestamp *time.Time) ([]common.MaintenanceMessage, error) {
+	// Truncate time to avoid updating again within seconds
+	if timestamp != nil {
+		if timestamp.After(time.Now()) {
+			return nil, merry.New("invalid time (future)")
+		}
+		truncated := timestamp.Truncate(time.Second * truncateTimeInterval)
+		timestamp = &truncated
+	}
+	if messages, ok := getMessagesFromCache(timestamp); ok {
+		return messages, nil
+	}
+
+	// Locking or waiting for another locked thread to complete
+	messageLock.Lock()
+	defer messageLock.Unlock()
```

Comment:
It feels like this will force all requests to be processed in a series. We don't really want that, do we?
I'm thinking something like this would make more sense:

```golang
// Global
lock := sync.RWMutex{}

// Local
lock.RLock()
	if messages, ok := getMessagesFromCache(timestamp); ok {
	        lock.RUnlock()
		return messages, nil
	}

// There is a small but non-zero chances that there are 2 threads in this section now
if lock.TryLock() {
messages, err := r.Queries.GetMaintenanceMessages(ctx)
	if err != nil {
		return nil, err
	}

	messageCache.Set("maintenance", timedCacheEntry{
		Cached: time.Now().Truncate(time.Second * truncateTimeInterval),
		Entry:  messages,
	}, cache.WithExpiration(time.Minute*10))
} else {
// Read from cache when it becomes available
lock.RLock()
	if messages, ok := getMessagesFromCache(timestamp); ok {
	        lock.RUnlock()
		return messages, nil
	} else {
 // ERROR
}
}
```